tracer/src/Datadog.Trace/SpanContextPropagator.cs

```diff
@@ -93,79 +72,61 @@ public void Inject<T>(SpanContext context, T carrier, Action<T, string, string>
 
             if (samplingPriority != null)
             {
-                setter(carrier, HttpHeaderNames.SamplingPriority, samplingPriority?.ToString(InvariantCulture));
+                setter(carrier, HttpHeaderNames.SamplingPriority, samplingPriority.Value.ToString(_invariantCulture));
             }
         }
 
         /// <summary>
         /// Extracts a <see cref="SpanContext"/> from the values found in the specified headers.
         /// </summary>
         /// <param name="headers">The headers that contain the values to be extracted.</param>
-        /// <typeparam name="T">Type of header collection</typeparam>
+        /// <typeparam name="TCarrier">Type of header collection</typeparam>
         /// <returns>A new <see cref="SpanContext"/> that contains the values obtained from <paramref name="headers"/>.</returns>
-        public SpanContext Extract<T>(T headers)
-            where T : IHeadersCollection
+        public SpanContext? Extract<TCarrier>(TCarrier headers)
+            where TCarrier : IHeadersCollection
         {
-            if (headers == null)
-            {
-                ThrowHelper.ThrowArgumentNullException(nameof(headers));
-            }
-
-            var traceId = ParseUInt64(headers, HttpHeaderNames.TraceId);
-
-            if (traceId == 0)
-            {
-                // a valid traceId is required to use distributed tracing
-                return null;
-            }
-
-            var parentId = ParseUInt64(headers, HttpHeaderNames.ParentId);
-            var samplingPriority = ParseSamplingPriority(headers, HttpHeaderNames.SamplingPriority);
-            var origin = ParseString(headers, HttpHeaderNames.Origin);
-
-            return new SpanContext(traceId, parentId, samplingPriority, null, origin);
+            return Extract(headers, DelegateCache<TCarrier>.Getter);
         }
 
         /// <summary>
         /// Extracts a <see cref="SpanContext"/> from the values found in the specified headers.
         /// </summary>
         /// <param name="carrier">The headers that contain the values to be extracted.</param>
         /// <param name="getter">The function that can extract a list of values for a given header name.</param>
-        /// <typeparam name="T">Type of header collection</typeparam>
+        /// <typeparam name="TCarrier">Type of header collection</typeparam>
         /// <returns>A new <see cref="SpanContext"/> that contains the values obtained from <paramref name="carrier"/>.</returns>
-        public SpanContext Extract<T>(T carrier, Func<T, string, IEnumerable<string>> getter)
+        public SpanContext? Extract<TCarrier>(TCarrier carrier, Func<TCarrier, string, IEnumerable<string?>> getter)
```

Comment:
Could you make sure to wait merge until after the following PR is merged https://github.com/DataDog/dd-trace-dotnet/pull/2202 ? I want to make sure that using nullable types like `string?` do not add unwanted assembly references, since AFAIK this is the first time we're introducing nullable reference types into the library

Reply:
`ServiceRemotingService.cs` has had nullable enable for months now :) It does not add any references, afaik. It's 100% a compile-time feature.

https://github.com/DataDog/dd-trace-dotnet/blob/master/tracer/src/Datadog.Trace/ServiceFabric/ServiceRemotingService.cs

But I don't mind waiting.

Reply:
Ah okay, I thought it was the first entry and I wanted to be super cautious about potential assembly references leaking in but TIL it does not 😄 Feel free to ignore this comment then!

Reply:
Thank you for being super cautious!